# Overview

An automation that pre-indexes datasets whenever one of the following events happens:

1) A new version of the dataset is downloaded
2) A new version of the alignment tool is installed

## Input Datasets Path
The input is stored by source-version/organism-dataset under the /data/scratch directory in a structure similar to that of /data/external/
For example, ensembl 91 genemone and transcriptome datasets for can be found under:
```
   Ensembl datasets (dna,cdna,pep,cdna,ncrna) for human:
       1) /data/scratch/ensembl-91/homo_sapiens-dna/
       2) /data/scratch/ensembl-91/homo_sapiens-cdna/
       3) /data/scratch/ensembl-91/homo_sapiens-ncrna/
       4) /data/scratch/ensembl-91/homo_sapiens-pep/
       5) /data/scratch/ensembl-91/homo_sapiens-cds/
       
```
This layout makes it easy to write an automation that generates the above structure whenever
the external source releases a new version.

Using the current_release_number file found in the root directory of each external source, we can check
whether or not the version in the file has been indexed by testing the existance of /data/scratch/source-version .


## Datasets Index Path

The indexes are generated by tool-version/source-version/  under /data/transformed. 

For example, ensembl 91 genemone indexes generated using bwa version v0.7.17 can be found
under:

```
   /data/transformed/bwa-v0.7.17/ensembl-91/homo_sapiens-dna/
```

In the case of ensembl datasets, the names of the organism(scientific name) and the dataset are part of the file name.

```
For example:
   Homo_sapiens.GRCh38.dna.chromosome.21.fa
```

Again this layout makes it easy for automation since each time we install a new version of the tool,
we update the content of the current_release_number file found in the root directory of each external tool

## Index File Format

Each index file is prefixed by tool_name-data_source_name-dataset_name 

Question: Should we also include the reference file name in the prefix?

## Testing Indexes?
How do we test that the indexes were generated properlly?


## Alingment Tools
* Bwa  -- http://bio-bwa.sourceforge.net/bwa.shtml 
```
Usage: $path2/bwa  index <reference_fasta>

Notes:
This gives .pac, .bwt, .ann, .amb and .sa index files that all have the same reference_fasta basename. 
Tools recognize index files within the same directory by their identical basename.

See: http://www.htslib.org/workflow/#mapping_to_variant
     https://github.com/lh3/bwa#type
   
```
* Bowtie -- http://bowtie-bio.sourceforge.net/manual.shtml
```
Usage: $path2/bowtie-build  <reference_fasta> <index_prefix>
See: http://bowtie-bio.sourceforge.net/tutorial.shtml#newi

```
* Bowtie2 -- http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml
```
Usage: $path2/bowtie2-build [options]* <reference_in> <bt2_base>
Where:
<reference_in>
A comma-separated list of FASTA files containing the reference sequences to be aligned to,
or, if -c is specified, the sequences themselves. E.g., <reference_in> might be chr1.fa,chr2.fa,
chrX.fa,chrY.fa, or, if -c is specified, this might be GGTCATCCT,ACGGGTCGT,CCGTTCTATGCGGCTTA.

<bt2_base>
The basename of the index files to write. By default, bowtie2-build writes files named NAME.1.bt2,
NAME.2.bt2, NAME.3.bt2, NAME.4.bt2, NAME.rev.1.bt2, and NAME.rev.2.bt2, where NAME is <bt2_base>.

See: http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer
```

* Hisat2  -- http://ccb.jhu.edu/software/hisat2/manual.shtml#the-hisat2-build-indexer
```
Usage: $path2/hisat2-build [options]* <reference_in> <ht2_base>

Notes:
If you use --snp, --ss, and/or --exon, hisat2-build will need about 200GB RAM for the human genome size 
as index building involves a graph construction.  Otherwise, you will be able to build an index 
on your desktop with 8GB RAM.

Main arguments:

<reference_in>
A comma-separated list of FASTA files containing the reference sequences to be aligned to, or, 
if -c is specified, the sequences themselves. E.g., <reference_in> might be chr1.fa,chr2.fa,chrX.fa,
chrY.fa, or, if -c is specified, this might be GGTCATCCT,ACGGGTCGT,CCGTTCTATGCGGCTTA.

<ht2_base>
The basename of the index files to write. By default, hisat2-build writes files named NAME.1.ht2, 
NAME.2.ht2, NAME.3.ht2, NAME.4.ht2, NAME.5.ht2, NAME.6.ht2, NAME.7.ht2, and NAME.8.ht2 
where NAME is <ht2_base>.
```
* kallisto. -- https://pachterlab.github.io/kallisto/starting

```
Usage: $path2/kallisto index [arguments] index_name.idx reference_in

Example: 
$path2/kallisto index -i kallisto-ensembl-cdna.idx /../ensembl-91/.../Danio_rerio.GRCz10.cdna.all.fa

Required argument:

-i, --index=STRING          Filename for the kallisto index to be constructed

Optional argument:

-k, --kmer-size=INT         k-mer (odd) length (default: 31, max value: 31)
    --make-unique           Replace repeated target names with unique names


The Fasta file supplied can be either in plaintext or gzipped format.

```


